tools/speech_data_explorer/data_explorer.py

```diff
@@ -1,846 +1,1044 @@
-def show_diff(idx, data):
-    if len(idx) == 0:
-        raise PreventUpdate
-
-    orig_words = data[idx[0]]['text']
-    orig_words = '\n'.join(orig_words.split()) + '\n'
-
-    pred_words = data[idx[0]]['pred_text']
-    pred_words = '\n'.join(pred_words.split()) + '\n'
-
-    diff = diff_match_patch.diff_match_patch()
-    diff.Diff_Timeout = 0
-    orig_enc, pred_enc, enc = diff.diff_linesToChars(orig_words, pred_words)
-    diffs = diff.diff_main(orig_enc, pred_enc, False)
-    diff.diff_charsToLines(diffs, enc)
-    diffs_post = []
-    for d in diffs:
-        diffs_post.append((d[0], d[1].replace('\n', ' ')))
-
-    diff_html = diff.diff_prettyHtml(diffs_post)
-
-    return diff_html
-
-
-@app.callback(Output('signal-graph', 'figure'), [Input('datatable', 'selected_rows'), Input('datatable', 'data')])
-def plot_signal(idx, data):
-    if len(idx) == 0:
-        raise PreventUpdate
-    figs = make_subplots(rows=2, cols=1, subplot_titles=('Waveform', 'Spectrogram'))
-    try:
-        filename = absolute_audio_filepath(data[idx[0]]['audio_filepath'], args.audio_base_path)
-        audio, fs = librosa.load(path=filename, sr=None)
-        if 'offset' in data[idx[0]]:
-            audio = audio[
-                int(data[idx[0]]['offset'] * fs) : int((data[idx[0]]['offset'] + data[idx[0]]['duration']) * fs)
-            ]
-        time_stride = 0.01
-        hop_length = int(fs * time_stride)
-        n_fft = 512
-        # linear scale spectrogram
-        s = librosa.stft(y=audio, n_fft=n_fft, hop_length=hop_length)
-        s_db = librosa.power_to_db(S=np.abs(s) ** 2, ref=np.max, top_db=100)
-        figs.add_trace(
-            go.Scatter(
-                x=np.arange(audio.shape[0]) / fs,
-                y=audio,
-                line={'color': 'green'},
-                name='Waveform',
-                hovertemplate='Time: %{x:.2f} s<br>Amplitude: %{y:.2f}<br><extra></extra>',
-            ),
-            row=1,
-            col=1,
-        )
-        figs.add_trace(
-            go.Heatmap(
-                z=s_db,
-                colorscale=[[0, 'rgb(30,62,62)'], [0.5, 'rgb(30,128,128)'], [1, 'rgb(30,255,30)'],],
-                colorbar=dict(yanchor='middle', lenmode='fraction', y=0.2, len=0.5, ticksuffix=' dB'),
-                dx=time_stride,
-                dy=fs / n_fft / 1000,
-                name='Spectrogram',
-                hovertemplate='Time: %{x:.2f} s<br>Frequency: %{y:.2f} kHz<br>Magnitude: %{z:.2f} dB<extra></extra>',
-            ),
-            row=2,
-            col=1,
-        )
-        figs.update_layout({'margin': dict(l=0, r=0, t=20, b=0, pad=0), 'height': 500})
-        figs.update_xaxes(title_text='Time, s', row=1, col=1)
-        figs.update_yaxes(title_text='Amplitude', row=1, col=1)
-        figs.update_xaxes(title_text='Time, s', row=2, col=1)
-        figs.update_yaxes(title_text='Frequency, kHz', row=2, col=1)
-    except Exception as ex:
```

Comment:
Please bring back the Exception handler (to log exception type)

Reply:
done